kimchi/src/circuits/gate.rs

```diff
@@ -342,25 +385,62 @@ impl GateType {
             chacha_final_where.insert((ChaChaFinal, *r));
         }
 
-        vec![
-            (chacha_pattern, chacha_where),
-            (chacha_final_pattern, chacha_final_where),
-        ]
+        let lookups = [
+            (chacha_pattern, chacha_where, Some(GateLookupTable::Xor)),
+            (
+                chacha_final_pattern,
+                chacha_final_where,
+                Some(GateLookupTable::Xor),
+            ),
+        ];
+
+        // Convert from an array of tuples to a tuple of vectors
+        {
+            let mut patterns = Vec::with_capacity(lookups.len());
+            let mut locations_with_tables = Vec::with_capacity(lookups.len());
+            for (pattern, locations, table) in lookups {
+                patterns.push(pattern);
+                locations_with_tables.push(GatesLookupSpec {
+                    gate_positions: locations,
+                    gate_lookup_table: table,
+                });
+            }
+            (patterns, locations_with_tables)
+        }
     }
 
-    pub fn lookup_kinds_map<F: Field>() -> HashMap<(GateType, CurrOrNext), usize> {
-        let mut res = HashMap::new();
-        let lookup_kinds = Self::lookup_kinds::<F>();
-        for (i, (_, locs)) in lookup_kinds.into_iter().enumerate() {
-            for (g, r) in locs {
-                if let std::collections::hash_map::Entry::Vacant(e) = res.entry((g, r)) {
+    #[allow(clippy::type_complexity)]
+    pub fn lookup_kinds_map<F: Field>(
+        locations_with_tables: Vec<GatesLookupSpec>,
+    ) -> GatesLookupMaps {
+        let mut index_map = HashMap::with_capacity(locations_with_tables.len());
+        let mut table_map = HashMap::with_capacity(locations_with_tables.len());
+        for (
+            i,
+            GatesLookupSpec {
+                gate_positions: locs,
+                gate_lookup_table: table_kind,
+            },
+        ) in locations_with_tables.into_iter().enumerate()
+        {
+            for location in locs {
+                if let std::collections::hash_map::Entry::Vacant(e) = index_map.entry(location) {
                     e.insert(i);
                 } else {
                     panic!("Multiple lookup patterns asserted on same row.")
                 }
+                if let Some(table_kind) = table_kind {
+                    if let std::collections::hash_map::Entry::Vacant(e) = table_map.entry(location)
+                    {
+                        e.insert(table_kind);
+                    }
+                }
             }
         }
-        res
+        GatesLookupMaps {
+            gate_selector_map: index_map,
```

Comment:
Index_map should be called gate_selector_map throughout, same for table_map

Reply:
We'd still need somewhere for the map to live if we want to use it in a closure. Might as well store it here directly.